filelog.go

```diff
@@ -345,6 +357,117 @@ func (w *FileLogWriter) archiveFiles(dir string) error {
 	return nil
 }
 
+func (w *FileLogWriter) compressFile(filename string, method string) {
+	basename := filepath.Base(filename)
+	dirname := filepath.Dir(filename)
+
+	plainFile, err := os.Open(filename)
+	if err != nil {
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't open logfile %q to begin compression: %s\n", w.filename, filename, err)
+		return
+	}
+
+	compressedFilename := filepath.Join(dirname, basename+".gz")
+	compressedInprogressFilename := compressedFilename + ".inprogress"
+
+	compressedFile, err := os.Create(compressedInprogressFilename)
+	if err != nil {
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't open new compressed file %q: %s\n", w.filename, compressedInprogressFilename, err)
+		return
+	}
+
+	// Defer closing the underlying file
+	defer func() {
+		err = compressedFile.Close()
+		if err != nil {
+			fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't close file %q: %s\n", w.filename, compressedInprogressFilename, err)
+		}
+	}()
+
+	var compressedFileWriter io.Writer
+	switch method {
+	case "gz":
+		compressedFileWriter = gzip.NewWriter(compressedFile)
+	case "zip":
+		// zipWriter is the outer container, compressedFileWriter is an entry inside the zip
+		zipWriter := zip.NewWriter(compressedFile)
+
+		// Defer closing the zip writer
+		defer func() {
+			err = zipWriter.Close()
+			if err != nil {
+				fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't close zip writer on %q: %s\n", w.filename, compressedInprogressFilename, err)
+			}
+		}()
+
+		compressedFileWriter, err = zipWriter.Create(basename)
+		if err != nil {
+			fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't open zip file entry: %s\n", w.filename, err)
+			return
+		}
+	default:
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Unknown compression method: %q\n", w.filename, method)
+		return
+	}
+
+	// Defer closing the writer, but only if we have a io.WriteCloser
```

Comment:
Sure, that's possible. In this case, I think it makes sense to leave where it is because `gz` follows the nominal structure, and `zip` is the outlier. If we added another compression algorithm, say `bzip2`, bzip2 would follow the same basic structure as `gz`, and the defer func() would handle both cases. Its unlikely we'd add support for `tar`, though.